# Basic Overview of Machine Learning Lifecycle

Machine Learning (ML) is the process of building systems that can learn patterns from data and make decisions or predictions.

The ML lifecycle is a step-by-step process that helps in building a successful ML model. It includes:

* Understanding the problem
* Collecting and exploring data
* Cleaning and preparing data
* Building and training the model
* Evaluating performance
* Making predictions and improving the model

This notebook gives a simple overview of these steps using an example dataset (Titanic dataset).

---

## Dataset Used

We are using the Titanic dataset available in seaborn

This dataset contains information about passengers such as age, gender, class, and whether they survived or not.

In [1]:
import seaborn as sns
df = sns.load_dataset("titanic")

---

## View First 5 Rows of Dataset

In [2]:
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


### Explanation

* `df.head()` is used to display the first 5 rows of the dataset.
* It helps to quickly understand the structure of the data.
* We can see column names, data types, and sample values.

### Output Understanding

* Each row represents one passenger.
* Each column represents a feature (like age, gender, ticket class, etc.).
* Example:

  * `survived`: 0 = No, 1 = Yes
  * `pclass`: Passenger class (1, 2, 3)
  * `sex`: Male or Female
  * `age`: Age of passenger
  * `fare`: Ticket price
---

## View Last 5 Rows of Dataset

In [3]:
df.tail()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
886,0,2,male,27.0,0,0,13.00,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.00,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.45,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.00,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.75,Q,Third,man,True,NaN,Queenstown,no,True


### Explanation

* `df.tail()` is used to display the last 5 rows of the dataset.
* It helps to check the data at the end of the dataset.
* Useful for verifying data loading and structure.

### Output Understanding

* Shows the last 5 passengers in the dataset.
* Structure is the same as `df.head()` (rows = passengers, columns = features).
* We can again observe values like:

  * `survived`: Survival status
  * `pclass`: Ticket class
  * `age`: Some values may be missing (`NaN`)
  * `embarked`: Port of boarding
---

## Check Shape of Dataset

In [4]:
df.shape

(891, 15)

### Explanation

* `df.shape` returns the size of the dataset.
* It shows the number of rows and columns.

### Output Understanding

* `(891, 15)` means:

  * **891 rows** → number of passengers
  * **15 columns** → number of features

This helps us understand how big the dataset is.

---

## View Column Names

In [5]:
df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

### Explanation

* `df.columns` is used to display all column names in the dataset.
* It helps to know what features are available.

### Output Understanding

* Shows a list of all 15 columns like:

  * `survived`, `pclass`, `sex`, `age`, `fare`, etc.
* These columns represent different features of passengers.
---

## Dataset Information

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


### Explanation

* `df.info()` gives a complete summary of the dataset.
* It shows:

  * Number of rows and columns
  * Column names
  * Data types of each column
  * Non-null (non-missing) values

### Output Understanding

* Total entries: **891 rows**

* Total columns: **15 columns**

* Data types:

  * `int64` → Integer values (e.g., survived, pclass)
  * `float64` → Decimal values (e.g., age, fare)
  * `object` → Text data (e.g., sex, embark_town)
  * `bool` → True/False values
  * `category` → Categorical data

* Missing values:

  * `age` has missing values (714 non-null)
  * `embarked` and `embark_town` have some missing values
  * `deck` has many missing values

This step helps in identifying **data types and missing values**, which is important for **data cleaning** in the ML lifecycle.

---

## Statistical Summary of Dataset

In [7]:
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### Explanation

* `df.describe()` gives statistical summary of all **numerical columns**.
* It helps to understand data distribution, central tendency, and spread.


### Output Understanding (How to Read)

Each column (like `age`, `fare`, etc.) has these values:

#### 1. **count**

* Number of non-missing values
* Example:

  * `age = 714` → 714 values present, rest are missing


#### 2. **mean (Average)**

* Sum of all values ÷ total values
* Example:

  * `age mean = 29.69` → average age is around 29 years


#### 3. **std (Standard Deviation)**

* Shows how spread out the data is from the mean
* High value → data is more spread
* Low value → data is closer to mean


#### 4. **min (Minimum)**

* Smallest value in the column
* Example:

  * `age min = 0.42` → youngest passenger


#### 5. **25% (1st Quartile)**

* 25% of data is below this value
* Example:

  * `age = 20.12` → 25% passengers are younger than ~20


#### 6. **50% (Median)**

* Middle value of the data
* 50% values are below and 50% are above
* Example:

  * `age = 28` → half passengers are younger than 28

👉 This is also called **median** (very important, especially when data has outliers)


#### 7. **75% (3rd Quartile)**

* 75% of data is below this value
* Example:

  * `age = 38` → most passengers are younger than 38


#### 8. **max (Maximum)**

* Largest value in the column
* Example:

  * `fare max = 512.32` → highest ticket price


### Key Insights from This Output

* `age` has missing values (count < 891)
* `fare` has high variation (large std) → possible outliers
* `sibsp` and `parch` mostly have low values (many zeros)
* Median (50%) is often better than mean when data is skewed


### Simple Understanding

* **mean** → average
* **50% (median)** → middle value
* **std** → spread of data
* **min/max** → range
* **25% & 75%** → distribution of data

---

## Check Missing Values

In [8]:
df.isnull().sum() 

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Explanation

* `df.isnull()` checks for missing values (NaN) in the dataset.
* `.sum()` adds them column-wise.
* So, this code shows total missing values in each column.

### Output Understanding

* Each value represents number of missing values in that column.

#### Important Observations:

* `age = 177`
  → 177 values are missing in age column

* `deck = 688`
  → Large number of missing values (most data is missing)

* `embarked = 2` and `embark_town = 2`
  → Very few missing values

* Other columns = 0
  → No missing values

### Simple Understanding

* **0** → No missing data
* **High number** → Many missing values (needs attention)


### Why This is Important

* Missing data can affect model performance
* We may need to:

  * Fill missing values (imputation)
  * Drop columns (like `deck` if too many missing)

---

## Check Duplicate Rows

In [9]:
df.duplicated().sum() 

np.int64(107)

### Explanation

* `df.duplicated()` checks for duplicate rows in the dataset.

* It returns **True/False** for each row:

  * `True` → duplicate row
  * `False` → unique row

* `.sum()` counts total duplicate rows.

### Output Understanding

* Output: **107**
  → There are 107 duplicate rows in the dataset

* `np.int64(107)` just means the number is stored as an integer type.


### Simple Understanding

* Duplicate rows = repeated data
* These can affect model performance

### Why This is Important

* Duplicate data can:

  * Bias the model
  * Give incorrect results

* Usually, we remove duplicates using:

---